src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs

```diff
@@ -132,7 +148,21 @@ public BodyModelBinder(IList<IInputFormatter> formatters, IHttpRequestStreamRead
                     return;
                 }
 
-                bindingContext.Result = ModelBindingResult.Success(model);
+                if (_isBindingRequired && model == null)
```

Comment:
Does (2) really satisfy the objective, though? I might be misjudging the goal, but I thought the point of this is that users don't want to have an extra explicit guard clause for `null` in their action - they want to be able to trust that `ModelState.IsValid` wouldn't be `true` if the `[FromBody]`-bound parameter was `null` (if they have to have a guard clause anyway, then they don't really need this change - they can just return `BadRequest` from it on their own). (2) on its own doesn't prevent the arrival of `null`, even in out-of-the-box cases with `JsonInputFormatter`.

> BTW current (1) approach misses cases like [FromBody] int count. The value seen in the input formatter when the body is empty will be a boxed 0, not null.

I agree that is a weakness. It would be better if we could really distinguish absent value-typed params. I was reasoning that it's not so bad, because we even in this case we satisfy the main objective that you definitely won't be getting `null` in your action method (you'd get `default(T)` instead, which in most cases is a legit input you'd have to allow for anyway). But if we could differentiate, I agree that would be a good thing.